sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala

```diff
@@ -0,0 +1,86 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.adaptive
+
+import org.apache.spark.MapOutputStatistics
+import org.apache.spark.sql.catalyst.plans.logical.{HintInfo, Join, JoinStrategyHint, LogicalPlan, NO_BROADCAST_HASH, PREFER_SHUFFLE_HASH, SHUFFLE_HASH}
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * This optimization rule includes two join selection:
+ *   1. detects a join child that has a high ratio of empty partitions and adds a
+ *      NO_BROADCAST_HASH hint to avoid it being broadcast.
+ *   2. detects a join child that every partition size less than local map threshold and adds a
+ *      PREFER_SHUFFLE_HASH hint to encourage being shuffle hash join instead of sort merge join.
+ *   3. if a join satisfies both NO_BROADCAST_HASH and PREFER_SHUFFLE_HASH,
+ *      then add a SHUFFLE_HASH hint.
+ */
+object DynamicJoinSelection extends Rule[LogicalPlan] {
+
+  private def shouldDemoteBroadcastHashJoin(mapStats: MapOutputStatistics): Boolean = {
+    val partitionCnt = mapStats.bytesByPartitionId.length
+    val nonZeroCnt = mapStats.bytesByPartitionId.count(_ > 0)
+    partitionCnt > 0 && nonZeroCnt > 0 &&
+      (nonZeroCnt * 1.0 / partitionCnt) < conf.nonEmptyPartitionRatioForBroadcastJoin
+  }
+
+  private def preferShuffledHashJoin(mapStats: MapOutputStatistics): Boolean = {
+    val localMapThreshold = conf.shuffleHashJoinLocalMapThreshold
+    mapStats.bytesByPartitionId.forall(_ <= localMapThreshold)
+  }
+
+  private def selectJoinStrategy(plan: LogicalPlan): Option[JoinStrategyHint] = plan match {
+    case LogicalQueryStage(_, stage: ShuffleQueryStageExec) if stage.resultOption.get().isDefined
+      && stage.mapStats.isDefined =>
+      val demoteBroadcastHash = shouldDemoteBroadcastHashJoin(stage.mapStats.get)
+      val preferShuffleHash = preferShuffledHashJoin(stage.mapStats.get)
+      if (demoteBroadcastHash && preferShuffleHash) {
+        Some(SHUFFLE_HASH)
+      } else if (demoteBroadcastHash) {
+        Some(NO_BROADCAST_HASH)
+      } else if (preferShuffleHash) {
+        Some(PREFER_SHUFFLE_HASH)
+      } else {
+        None
+      }
+
+    case _ => None
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformDown {
+    case j @ Join(left, right, _, _, hint) =>
+      var newHint = hint
+      if (!hint.leftHint.exists(_.strategy.isDefined)) {
+        selectJoinStrategy(left).foreach { strategy =>
+          newHint = newHint.copy(leftHint =
+            Some(hint.leftHint.getOrElse(HintInfo()).copy(strategy = Some(strategy))))
+        }
+      }
+      if (!hint.rightHint.exists(_.strategy.isDefined)) {
+        selectJoinStrategy(right).foreach { strategy =>
+          newHint = newHint.copy(rightHint =
+            Some(hint.rightHint.getOrElse(HintInfo()).copy(strategy = Some(strategy))))
+        }
+      }
```

Comment:
It seems possible that both sides have `PREFER_SHUFFLE_HASH` hint. How do we decide the build side?

Reply:
The same question applies to `SHUFFLE_HASH` as well, but that's an existing problem.

Reply:
e.g. `SHUFFLE_HASH(t1, t2)` with an inner join, do we always pick the left side as the build side, even if the right side is smaller?

Reply:
Ah sorry I misread the code. At the end we call `getBuildSide`, and pick the smaller side as the build side, if both sides are be the build side.

Reply:
yea, and has already added a test before [c747a23](https://github.com/apache/spark/pull/32550/commits/c747a23be0e59bcb7edb6e9bb377d2aed7248c96)